### PR TITLE
[DependencyInjection] Added "ignore" option to "_instanceof" definitions

### PR DESCRIPTION
**Pull request by tsantos84**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      |no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | awaiting the approval of this PR

As discussed on #35581, this PR aims to add the capability of ignore some implementations when configuring definitions with `_instanceof` option.

### Problem

The Dependency Injection component has a convenient way to create default configuration for services that implementing some interface or extending some classe. Although we can add those default configurations, we can't skip some implementations in this process.

Consider the following PHP classes and interface:

```php
interface FooInterface {}
class Foo implements FooInterface {}
class Bar implements FooInterface {}
class Baz implements FooInterface {}
```

If you want, for example, to not have `Baz` service configured from `_instanceof` configuration, you can change your service files like this:

### YAML file:

```yaml
services:
    _instanceof:
        FooInterface:
            tags: [{name:"foo"}]
            ignore:
                - Baz
            # or simple string for only one ignored implementation
            ignore: 'Baz'
    Foo: ~
    Bar: ~
    Baz:
        tags: [{name:"app.baz"}]
```

### XML file
```xml
<?xml version="1.0" encoding="utf-8"?>
<container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services https://symfony.com/schema/dic/services/services-1.0.xsd">
    <services>
        <instanceof id="FooInterface">
            <tag name="foo" />
            <ignore name="Baz" />
        </instanceof>
        <service id="Foo" />
        <service id="Bar" />
        <service id="Baz">
            <tag name="app.baz" />
        </service>
    </services>
</container>
```

### Remarks

* In the original issue, was proposed to add an `exclude` option but personally I think that `ignore` is more concise for this feature, although the component has already the option `exclude` to skip files and directories when configuring resources.
* Ignoring a implementation is valid per file only.

### CODE REVIEW

File: CHANGELOG-5.1.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+CHANGELOG for 5.1.x
```
Comment (nicolas-grekas) [inline CHANGELOG-5.1.md]: this file should be removed: the main changelog is autogenerated
instead, please update the changelog of the component

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 2]
```diff
@@ -147,7 +148,14 @@ protected function setDefinition($id, Definition $definition)
             }
             $this->instanceof[$id] = $definition;
         } else {
-            $this->container->setDefinition($id, $definition instanceof ChildDefinition ? $definition : $definition->setInstanceofConditionals($this->instanceof));
+            $conditionals = array_filter($this->instanceof, function (string $instanceOfId) use ($id) {
+                if (!isset($this->instanceofIgnored[$instanceOfId])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: no need for this "if" block, the next condition is enough as it embeds this check already

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/services_instanceof_ignore_class.xml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services https://symfony.com/schema/dic/services/services-1.0.xsd">
+    <services>
+        <instanceof id="Symfony\Component\DependencyInjection\Tests\Fixtures\BarInterface" lazy="true" autowire="true">
+            <tag name="foo" />
+            <ignore name="Symfony\Component\DependencyInjection\Tests\Fixtures\Baz" />
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/services_instanceof_ignore_class.xml]: `<ignore>Symfony\Component\DependencyInjection\Tests\Fixtures\Baz</ignore>`?

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services_instanceof_invalid_ignore.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+services:
+    _instanceof:
+        Symfony\Component\DependencyInjection\Tests\Fixtures\BarInterface:
+            autowire: true
+            lazy: true
+            tags:
+                - { name: foo }
+            ignore: 'Symfony\Component\DependencyInjection\Tests\Fixtures\Baz'
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services_instanceof_invalid_ignore.yml]: we could make it work actually, that might be useful

**Review by nicolas-grekas — COMMENTED** [review]
thanks, here are some comments on the implementation
don't miss patching `InstanceofConfigurator` also

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services_instanceof_invalid_ignore.yml (hunk 4 above)
Comment (tsantos84) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services_instanceof_invalid_ignore.yml]: I've thought about this option, but honestly I don't have the sensibility if this is could grows the learning curve or not. In this case, I'm with your feeling. :)

File: CHANGELOG-5.1.md (hunk 1 above)
Comment (tsantos84) [reply CHANGELOG-5.1.md]: Done.

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php (hunk 2 above)
Comment (tsantos84) [reply src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: I was wondering whether `isset($this->instanceofIgnored[$instanceOfId])` could throw a warning if `$instanceOfId` is not presented in the array or not. Done!

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/services_instanceof_ignore_class.xml (hunk 3 above)
Comment (tsantos84) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/services_instanceof_ignore_class.xml]: I've followed the same signature of tags but I agree that your version is better.

**Review by nicolas-grekas — DISMISSED** [review]
The implementation looks good to me.
Feature-wise, this is an edge case, but I'm no objections against supporting it - this has already been reported before.

**Review by nicolas-grekas — COMMENTED** [review]
I just realized that this implementation ignores by "id" and not by "fqcn".
Maybe that's for the best, but this needs to be a conscious decision and the wording should be updated accordingly.
About "ignore" vs "exclude", I'm not sure about "ignore" anymore: we started the discussion with "exclude", which could mean that's the most expected word. For thoughts.

**Comment by Tobion** [comment]
-1 as there is a clear solution with separate files (https://github.com/symfony/symfony/issues/35581#issuecomment-581943325) which does not need documentation or explanation what `ignore` supports and what it doesn't.

**Comment by tsantos84** [comment]
@nicolas-grekas 
> About "ignore" vs "exclude", I'm not sure about "ignore" anymore: we started the discussion with "exclude", which could mean that's the most expected word. For thoughts.

Maybe "exclude" is a better term because devs already knows what "exclude" means in this context.

@Tobion 
> -1 as there is a clear solution with separate files (#35581 (comment)) which does not need documentation or explanation what ignore supports and what it doesn't.

I agree with you that it'll work. I've just create a demo application to see how easy it'd be to accomplish the same feature by creating two separated files. The order of loading the file matters. Depending on how I load the new file in Kernel (before or after "services.yaml") and configure those services, the "exclude" feature will not work. So, this is not so easy to understand, mainly for newcomers. 

Example with two separated files:

```yaml
#services.yaml
services:
    _instanceof:
        App\Foo\FooInterface:
            tags: [{name: "app.foo"}]

    App\Foo\FooInterface: '@App\Foo\FooComposite'

#services_foo.yaml
services:
    App\Foo\FooComposite:
        arguments: [!tagged_iterator "app.foo"]
```

With these files, I'm obligated to load `services_composite.yaml` after `services.yaml`. Another point is that creating two separated files to implement a simple design pattern is not straightforward, IHMO.

**Comment by Tobion** [comment]
> With these files, I'm obligated to load services_composite.yaml after services.yaml.

What's the reason?

**Comment by tsantos84** [comment]
> What's the reason?

The following error is raised:

`Cannot autowire service "App\Foo\FooComposite": argument "$foos" of method "__construct()" is type-hinted "iterable", you should configure its value explicitly.`

**Comment by Tobion** [comment]
Then you probably still load FooComposite inside services.yaml which overwrites services_composite.

**Comment by nicolas-grekas** [comment]
I think this discussion is a good justification to add the feature :)

**Comment by Tobion** [comment]
The service file is just not completely given. So it's impossible to tell what they are doing wrong.
If you start adding exlusion rules for certain rules, then you need to add exclusions for all of them. What about `bind` etc? They all have the same logic: Services loaded/defined from this file.

**Comment by tsantos84** [comment]
> The service file is just not completely given. So it's impossible to tell what they are doing wrong.

I've just pushed the demo application on https://github.com/tsantos84/di-composite-example. The important files are:

1. config/services.yaml
2. config/services_foo.yaml
3. src/Foo/*
4. src/Controller/

> If you start adding exlusion rules for certain rules, then you need to add exclusions for all of them. What about `bind` etc? They all have the same logic: It applies to services loaded/defined from this file.

I believe when a tool allows me to configure something in batch mode (`_instanceof` in this case), then I expect that tool gives me the possibility to "exclude/ignore" part of the content being handled. `_defaults`, `bind` and etc are very straightforward because they can be easily redefined on a specific service definition, and this is not the case for `_instanceof` feature.

**Comment by Tobion** [comment]
Thanks for the demo.

https://github.com/tsantos84/di-composite-example/blob/495a21a67ab8b925720dc3cbf9dc455a089d3401/config/services.yaml#L16-L18
includes the FooComposite. That means it tries to autowire it and it tries to apply _instanceof on it. You need to exclude it as I already assummed

> Then you probably still load FooComposite inside services.yaml which overwrites services_composite.

**Comment by tsantos84** [comment]
> includes the FooComposite. That means it tries to autowire it and it tries to apply _instanceof on it. You need to exclude it as I already assummed

Finally got the final version to get the solution done using two separated files:

```yaml
#services.yaml
services:
  _instanceof:
    App\Foo\FooInterface:
      tags: [{name: "app.foo"}]

  App\:
    resource: '../src/*'
    exclude: '../src/{DependencyInjection,Entity,Migrations,Tests,Kernel.php,Foo/FooComposite.php}'

#services_foo.yaml
services:
  App\Foo\FooComposite:
    arguments: [!tagged_iterator "app.foo"]

  App\Foo\FooInterface: '@App\Foo\FooComposite'
```

As I told before, it will work splitting the definition in two separated files, but is not so simple to accomplish as we need to exclude totally FooComposite definition from `services.yaml` and define it on its own yaml file (and make reference to tags defined in `services.yaml` file).

**Comment by tsantos84** [comment]
After discussing this proposal with @Tobion here and @nicolas-grekas on Slack, I prefer to stop this contribution as there are few use cases that will be impacted with this solution even they exist. If some day more people ask for this, I think it will be reconsidered to make part of base code. Until that, solving the problem on dev side would be a better solution. Thanks @Tobion and @nicolas-grekas for the rich discussion on this topic.

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/services_instanceof_ignore_class.xml (hunk 3 above)
Comment (stof) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/services_instanceof_ignore_class.xml]: tags allow having arbitrary attributes

**Pull request closed without merging**